### PR TITLE
Procedures check security

### LINKED ISSUE #227: Security: ProviderSDK needs to make an auth check to papiea when executing procedures
When the providerSdk is receiving a procedure call, it gets the `headers` from papiea (needs #225). It then needs to make a call back to papiea, to ensure that the received token is in-fact allowed to execute this function. Otherwise, some third party may be able to send commands to run procedures on providers without properly authenticating.

### PR DESCRIPTION
**Pull request by Emulebest**
resolves #227

### CODE REVIEW

File: papiea-engine/src/auth/authz.ts [hunk 1]
```diff
@@ -9,31 +9,21 @@ export class PermissionDeniedError extends Error {
...
-    private action: string;
-
-    constructor(action: string) {
-        this.action = action;
-    }
-
-    getAction(): string {
-        return this.action;
-    }
+export enum Actions {
```
Comment (adolgarev) [inline papiea-engine/src/auth/authz.ts]: what for changed class to Enum? Anyways do not make it plural, it serves as a type and  declaration `action : Actions` is not right as it says that argument action represents many actions, it should be `action : Action`
BTW rename then ReadAction to Read, etc

File: papiea-engine/src/entity/entity_routes.ts [hunk 2]
```diff
@@ -31,6 +32,25 @@ export function createEntityAPIRouter(entity_api: Entity_API): Router {
         return {results: pageEntities, entity_count: totalEntities};
     };
 
+    router.post("/:prefix/:version/check_permission", asyncHandler(async (req, res) => {
+        const allowed: boolean = await entity_api.check_permission(req.user, req.params.prefix, req.params.version, req.body.action, req.body);
```
Comment (adolgarev) [inline papiea-engine/src/entity/entity_routes.ts]: api call MUST check permission, the check here is redundant

File: papiea-sdk/src/provider_sdk/typescript_sdk_context_impl.ts [hunk 3]
```diff
@@ -52,14 +89,28 @@ export class ProceduralCtx implements ProceduralCtx_Interface {
...
-            if (parts[0] === 'Bearer') 
+            const parts = this.headers.authorization.split(' ');
+            if (parts[0] === 'Bearer')
                 return parts[1]
         }
         throw new Error("No invoking user")
     }
+}
+
+export enum Actions {
```
Comment (adolgarev) [inline papiea-sdk/src/provider_sdk/typescript_sdk_context_impl.ts]: why not to move this to core then and do not duplicate the code?

File: papiea-engine/src/entity/entity_api_interface.ts [hunk 4]
```diff
@@ -22,4 +24,8 @@ export interface Entity_API {
     call_kind_procedure(user: UserAuthInfo, prefix: string, kind_name: string, version: Version, procedure_name: string, input: any): Promise<any>
 
     call_provider_procedure(user: UserAuthInfo, prefix: string, version: Version, procedure_name: string, input: any): Promise<any>
+
+    check_permission(user: UserAuthInfo, prefix: string, version: Version, action: Actions, entityRef: Entity_Reference): Promise<boolean>
+
+    check_permissions(user: UserAuthInfo, prefix: string, version: Version, action: Actions, requestParams: any): Promise<boolean>
```
Comment (adolgarev) [inline papiea-engine/src/entity/entity_api_interface.ts]: what is requestParams?

File: papiea-engine/src/auth/authz.ts (hunk 1 above)
Comment (Emulebest) [reply papiea-engine/src/auth/authz.ts]: If this a class, we can't transfer this as string param. I am not sure we had this as a class in the first place

File: papiea-engine/src/entity/entity_api_interface.ts (hunk 4 above)
Comment (Emulebest) [reply papiea-engine/src/entity/entity_api_interface.ts]: either `metadata` or `entity_ref`

Comment (shlomi-nx) [reply papiea-engine/src/entity/entity_api_interface.ts]: `Metadata` is an `Entity_Reference` by implementation. If anything, this type could have been `Metadata[]`. However, I think this actually needs to be a list of `[Action, Metadata]` tuple, this way, the singular case is just a degenerate form where the list is of size 1.

File: papiea-engine/src/entity/entity_routes.ts (hunk 2 above)
Comment (shlomi-nx) [reply papiea-engine/src/entity/entity_routes.ts]: @adolgarev Did not quite get your message, this check is exactly what this api is doing - checking that these operations are allowed.

File: papiea-engine/src/entity/entity_routes.ts [hunk 5]
```diff
@@ -31,6 +32,25 @@ export function createEntityAPIRouter(entity_api: Entity_API): Router {
...
+    }));
+
+    router.post("/:prefix/:version/check_permissions", asyncHandler(async (req, res) => {
+        const allowed: boolean = await entity_api.check_permissions(req.user, req.params.prefix, req.params.version, req.body.action, req.body);
+        if (allowed) {
+            res.json({"success": "Ok"})
+        } else {
+            throw new PermissionDeniedError()
+        }
+        console.log()
```
Comment (shlomi-nx) [inline papiea-engine/src/entity/entity_routes.ts]: Why is this here?

File: papiea-sdk/src/provider_sdk/typescript_sdk.ts [hunk 6]
```diff
@@ -22,50 +22,50 @@ class SecurityApiImpl implements SecurityApi {
     readonly provider: ProviderSdk;
     readonly s2s_key: string;
     constructor (provider:ProviderSdk, s2s_key:Key) {
-        this.provider = provider
+        this.provider = provider;
```
Comment (shlomi-nx) [inline papiea-sdk/src/provider_sdk/typescript_sdk.ts]: Just as a side note, I see you keep adding `;` to the end of lines. I personally try to avoid that, I do not like that style. We have gone a long way from c ;) Don't feel compelled to put these everywhere, actually, you are encouraged to remove them.

File: papiea-sdk/src/provider_sdk/typescript_sdk.ts [hunk 7]
```diff
@@ -410,6 +410,7 @@ export class Kind_Builder {
         const version = this.get_version();
         this.server_manager.register_handler(`/${this.kind.name}/${name}`, async (req, res) => {
             try {
+                console.log(req.headers);
```
Comment (shlomi-nx) [inline papiea-sdk/src/provider_sdk/typescript_sdk.ts]: Please add contextual messages to log prints

File: papiea-sdk/src/provider_sdk/typescript_sdk_utils.ts [hunk 8]
```diff
@@ -1,3 +1,6 @@
+import { Metadata } from "papiea-core";
+import { Entity_Reference } from "papiea-core/build/core";
```
Comment (shlomi-nx) [inline papiea-sdk/src/provider_sdk/typescript_sdk_utils.ts]: wrong `from` clause. This should just be added to the line above.

File: papiea-sdk/src/provider_sdk/typescript_sdk_interface.ts [hunk 9]
```diff
@@ -63,6 +65,7 @@ export interface IntentfulCtx_Interface {
     get_user_security_api(user_s2skey:Key): SecurityApi
     get_headers(): IncomingHttpHeaders
     get_invoking_token(): Key
+    check_permission(provider_prefix: string, provider_version: Version, entity: any, action: Actions): Promise<boolean>
```
Comment (shlomi-nx) [inline papiea-sdk/src/provider_sdk/typescript_sdk_interface.ts]: please remove `prefix` and `version`. They are to be picked up from the context. 

If we will want to add the ability to check permissions for other providers we will add that explicitly.

File: papiea-sdk/src/provider_sdk/typescript_sdk.ts [hunk 10]
```diff
@@ -454,6 +454,7 @@ export class Kind_Builder {
...
```
Comment (shlomi-nx) [inline papiea-sdk/src/provider_sdk/typescript_sdk.ts]: @Emulebest Please stop pushing logs with no strings for context.

Comment (Emulebest) [reply papiea-sdk/src/provider_sdk/typescript_sdk.ts]: Sorry, will double check for debug logs
